### PR TITLE
Better error message

### PR DESCRIPTION
**Pull request by nouiz**
It is to help debug a corner case on Windows found by in pymc3

### CODE REVIEW

**Comment by nouiz** [comment]
do not merge yet.

**Comment by nouiz** [comment]
I think it is ready to review.

**Comment by lamblin** [comment]
LGTM.
OK to merge when the pymc3 user confirms it prints the right thing for them.

File: theano/gof/link.py [hunk 1]
```diff
@@ -157,6 +157,9 @@ def raise_with_op(node, thunk=None, exc_info=None, storage_map=None):
         detailed_err_msg += ("Inputs shapes: %s" % shapes +
                              "\nInputs strides: %s" % strides +
                              "\nInputs values: %s" % scalar_values)
+        if theano.config.exception_verbosity == 'high':
+            detailed_err_msg += "\nInputs type_num: %s" % str(
+                [getattr(getattr(i[0], 'dtype', ''), 'num', '') for i in thunk.inputs])
```
Comment (lamblin) [inline theano/gof/link.py]: Not working apparently, since it prints `''` each time.

File: theano/tensor/subtensor.py [hunk 2]
```diff
@@ -1934,7 +1934,7 @@ def copy_of_x(self, x):
         # max_depth: we pass 0 to have this parameter ignored
         # requirements: here we pass NPY_ARRAY_ENSURECOPY to force a copy
         # context: this is almost always NULL, I'm not sure what it's used for
-        return """(PyArrayObject*)PyArray_FromAny(py_%(x)s, NULL, 0, 0,
+        return """(PyArrayObject*)PyArray_FromAny(py_%(x)s, PyArray_DTYPE(%(x)s), 0, 0,
```
Comment (abergeron) [inline theano/tensor/subtensor.py]: Maybe the comment above needs to be updated.

**Comment by nouiz** [comment]
I updated this PR.

**Comment by abergeron** [comment]
Ok with me, I'll let @lamblin see if his concerns are addressed.

**Comment by nouiz** [comment]
It don't always works. In the user case we wanted to debug, it wasn't
working, by when I test manually it work.

So as it work some times, I would merge this.

I was not able to reproduce the not working case.

**Comment by lamblin** [comment]
It makes a better error message, and sometimes it helps, so I'm OK with merging.
It would be nice to have a way of reproducing the issue, though...

File: theano/tensor/subtensor.py (hunk 2 above)
Comment (lamblin) [inline theano/tensor/subtensor.py]: According to the doc, `PyArray_DTYPE` returns a borrowed reference, and `PyArray_FromAny` steals it, so we need to call `Py_INCREF` on that dtype.

Comment (nouiz) [inline theano/tensor/subtensor.py]: This is complicated. As this was not confirmed to help on windows, I'll revert it.

Comment (nouiz) [inline theano/tensor/subtensor.py]: https://github.com/Theano/Theano/pull/4606

**Pull request merged**